Add documentation for schemas

cari-lynn opened the pull request:
Related to this there will be another PR to add examples of using schema to the `examples/` directory in the `ytt` repo.

cari-lynn committed:
Add documentation for schemas

- most of the docs were largely taken from the schema spec that exists
  in hackmd
- document scalers `nullable` and `#@schema/type any=True`

pivotaljohn commented on content/ytt/docs/latest/lang-ref-ytt-schema.md:
```diff
@@ -0,0 +1,248 @@
+
+## Overview
+
+`ytt`'s Schema feature provides a way to verify YAML structures and values with the help of annotations.
+Schema currently allows you to provide default data values, and verify that each data value file provided is compatible with the schema.
+
+The schema definition spells out the set of nodes that must make up that kind of YAML document. Documents that are type checked and validated against the schema are guaranteed to be proper.
+
+For example, in the case of Data Values, the schema guarantees to templates that all values exist and are of the correct type. The aleviates templates from doing existence and type checks, themselves.
```
For this release of the Schema feature, I think it keeps the message more concise/clear if we describe it as being _all about_ Data Values. It's gonna be some time until we support checking a document via schema, generally.

pivotaljohn commented on content/ytt/docs/latest/lang-ref-ytt-schema.md:
```diff
@@ -0,0 +1,248 @@
+
+For example, in the case of Data Values, the schema guarantees to templates that all values exist and are of the correct type. The aleviates templates from doing existence and type checks, themselves.
+
+There are two features to verify a document is compatible with a schema:
+- type checking — comparing the shape of a YAML structure against its schema’s types.
+- validation — determining whether the values contained in a YAML structure are acceptable.
```
We're not supporting validation in this release. I think we can nix this paragraph.

pivotaljohn commented on content/ytt/docs/latest/lang-ref-ytt-schema.md:
```diff
@@ -0,0 +1,248 @@
+There are two features to verify a document is compatible with a schema:
+- type checking — comparing the shape of a YAML structure against its schema’s types.
+- validation — determining whether the values contained in a YAML structure are acceptable.
+---
```
Provide a simple example, right here?  (or write one elsewhere and link to it from here?)

This will create a personal first exposure/experience on which to hang the concepts that follow.

cari-lynn replied:
I moved up the section on how to define schema so that readers can get a quick introduction into the syntax, but I would also like to add a link to an example here.

pivotaljohn commented on content/ytt/docs/latest/lang-ref-ytt-schema.md:
```diff
@@ -0,0 +1,248 @@
+- type checking — comparing the shape of a YAML structure against its schema’s types.
+- validation — determining whether the values contained in a YAML structure are acceptable.
+---
+## Supported Types
+
+While YAML provides for an extendable range of types, the `ytt` Schema supports a specific set.
+
+### Scalar Types
+
+- `bool` — `true`, `false` (and when not strict, `yes`, `no`, `Y`, `N`, etc.)
+- `float` — e.g. `0.4`
+- `int` — e.g. `42`
+- `null` — `null`, `~`, and when value is omitted.
+- `string` — e.g. `""`, `"ConfigMap"`, `"0xbeadcafe"`
+
+
+### Algebraic Types
+
+- `any` — any valid YAML value is permitted.
+
+### Collection Types
+
+A collection type is the pair of itself (either `map` or `array`) and its contained type.
+
+#### Array type
+
+`array<type>` — a YAML sequence, each item must be of the specified `type`.
+
+Examples:
+- `array<string>` — a sequence of strings (and only strings)
+- `array<map<...>>` — a sequence containing only map instances of the same shape.
+
+#### Map type
+
+`map <keyA:typeA, keyB:typeB, ...>` — a YAML mapping that must contain exactly one item of each given key, whose value must be of the type specified.
+
```
Let's simplify this:
- introduce as few concepts as possible (e.g. remove "Algebraic Types" as an idea and explain `any` a special type?
- I don't think anything is lost if we merely refer to "Arrays" and "Maps"... we've been somewhat consistent in referring to their direct children as "array item" and "map item"
- the angle-bracket notation isn't showing up in any interface the user sees and might be confusing to introduce it here (as if it's a notation they should be aware of).  Instead, perhaps give concrete examples?

cari-lynn replied:
I modified this section to remove algebraic types, and kept maps and arrays in their own sections but provided examples. I am torn between putting maps and arrays with the other types, or keeping them on their own.

pivotaljohn reviewed (commented):
Hey @cari-lynn, a partial review. Prolly helpful to go over together as some of the feedback is rather abstract.

cari-lynn commented:
Checklist of things to address:
- [ ] Can we assume readers know what configuration consumers/authors are and which they are?
- [ ] Add a real life example (playground?)

cari-lynn committed:
Simplify type section

cari-lynn committed:
Add "Using Schemas" section to docs

- it introduces schemas at a high level and provides an example
- updates the `lang-ref-schema.md` to simplify type section

pivotaljohn commented on content/ytt/docs/latest/ytt-schema.md:
```diff
@@ -0,0 +1,60 @@
+
+`ytt` schemas are currently in the **experimental** phase. To use schema features, include `--enable-experiment-schema`.
+
+Configuration authors use Schema to verify data value YAML structures and their values and define default data value configuration.
```
Mentioning persona is great. 👍🏼 

Thinking pedantically about what action the Configuration Author takes... they are responsible for _declaring_ the ~~variables~~ data values.

```suggestion
Configuration Authors use Schema to declare data values; specifying the type and default value for each.
```

pivotaljohn commented on content/ytt/docs/latest/ytt-schema.md:
```diff
@@ -0,0 +1,60 @@
+
+Configuration authors use Schema to verify data value YAML structures and their values and define default data value configuration.
+
+Using a schema guarantees to templates that all data values exist and are of the correct type. This alleviates templates from doing existence and type checks themselves.
```
🥇

pivotaljohn commented on content/ytt/docs/latest/ytt-schema.md:
```diff
@@ -0,0 +1,60 @@
+
+Using a schema guarantees to templates that all data values exist and are of the correct type. This alleviates templates from doing existence and type checks themselves.
+
+Common use cases:
+
+1. Set default values
+1. Provide visibility to valid values that could be updated by the consumer 
+1. Provide visibility to valid types and YAML structure that could be updated by the consumer
```
With my "new user" empathy cap on, I wonder if we've already motivated the core use case.

```suggestion
```

pivotaljohn commented on content/ytt/docs/latest/ytt-schema.md:
````diff
@@ -0,0 +1,60 @@
+>   enabled: true
+>   external_ip: ""
+> ```
+A configuration _consumer_ may add their own customized data values. The data values used are limited to only the values in the schema above, and if supplied, `load_balancer.enabled` must be a `bool`, and `load_balancer.external_ip` must be a `string`.
````
Yeah, "Data Values" is a bit of an unfortunate name because it's rather overloaded: it's the variables, it's also the values, it's also a file, it's also the result of a merge of a bunch of files. *sigh*

plus:
- I think it helps when we treat persona titles as if they are proper nouns... kinda helps them stand out.

```suggestion
A Configuration _Consumer_ may customize these by providing their own data values. If supplied, `load_balancer.enabled` must be a `bool`, and `load_balancer.external_ip` must be a `string`. No additional data values can be included.
```

cari-lynn replied:
Yes this writing is exactly what I was trying to do thank you for helping me get there!

pivotaljohn commented on content/ytt/docs/latest/ytt-schema.md:
````diff
@@ -0,0 +1,60 @@
+>   external_ip: 172.120.12.232
+> ```
+
+Then paired with a template
+> `config.yml`
+> ```yaml
+> #@load("ytt:data", "data")
+> ---
+> service: #@ data.values.load_balancer
+> ```
+> 
````
If you pair-up the template with the schema, it hits home that the Configuration Author writes both of _those_ files... and then in the next scene, the Configuration Consumer writes the `values.yml`.

cari-lynn replied:
Great point

pivotaljohn commented on content/ytt/docs/latest/ytt-schema.md:
```diff
@@ -0,0 +1,60 @@
+---
+title: ytt @schema/match
```
```suggestion
title: Using Schema
```

pivotaljohn reviewed (commented):
Really diggin' the introduction page. This review focuses on just that page.

pivotaljohn commented on content/ytt/docs/latest/lang-ref-ytt-schema.md:
```diff
@@ -0,0 +1,226 @@
+
+`ytt`'s Schema feature provides a way to verify data value YAML structures and values with the help of annotations.
+
+A schema allows you to provide default data values, and verify that each data value file provided is type checked and validated against the schema and guaranteed to be proper.
```
critique: compound sentence. I think it would help to enumerate, chronologically, how schema participate.

pivotaljohn commented on content/ytt/docs/latest/lang-ref-ytt-schema.md:
```diff
@@ -0,0 +1,226 @@
+  is not set to `True`, this results in an error.
+
+Notes:
+- Schema must be defined _before_ any other files containing YAML can be processed. Therefore, a file containing a Schema document must not contain other kinds of documents.
```
```suggestion
- a file containing a Schema document must not contain other kinds of documents. This is because schema must be defined _before_ templates can be processed.
```

- The most important bit of information for the user is the rule... and _then_ the rational.
- we can narrow the set of files to just "templates", here. They are the only files that can load data values. and I believe it helps to make it (slightly) more concrete.

pivotaljohn commented on content/ytt/docs/latest/lang-ref-ytt-schema.md:
```diff
@@ -0,0 +1,226 @@
+
+Notes:
+- Schema must be defined _before_ any other files containing YAML can be processed. Therefore, a file containing a Schema document must not contain other kinds of documents.
+- files containing Schema documents will be detected among input files (i.e. those specified by `-f`).
```
Ditto from comment above about putting the rule first.

pivotaljohn commented on content/ytt/docs/latest/lang-ref-ytt-schema.md:
```diff
@@ -0,0 +1,226 @@
+Notes:
+- Schema must be defined _before_ any other files containing YAML can be processed. Therefore, a file containing a Schema document must not contain other kinds of documents.
+- files containing Schema documents will be detected among input files (i.e. those specified by `-f`).
+- the first file containing a Schema document establishes the "base". Subsequent files containing Schema documents are overlayed onto the "base" in the order they appear (identically to how Data Values files are processed).
```
link to that description of the order in which Data Values are processed?

cari-lynn replied:
I ended up rewording this to use the same verbage as data values overlays docs.

pivotaljohn commented on content/ytt/docs/latest/lang-ref-ytt-schema.md:
```diff
@@ -0,0 +1,226 @@
+- `int` — e.g. `42`
+- `null` — `null`, `~`, and when value is omitted.
+- `string` — e.g. `""`, `"ConfigMap"`, `"0xbeadcafe"`
+- `any` — any valid YAML value is permitted. Must be set [explicitly](#schematype).
```
Might well pull this one apart from the others. Perhaps just at the bottom of the list? or called out as a special type?

cari-lynn replied:
I moved it to the bottom of the list. It is different from the other types, but I am not sure the reader gains any info that would help then write their schema  by separating it. Happy to hear other thoughts though

pivotaljohn replied:
I like it. Better this way, I think... and the link helps make clear how. 👍🏼

pivotaljohn commented on content/ytt/docs/latest/lang-ref-ytt-schema.md:
```diff
@@ -0,0 +1,226 @@
+---
+## Inferring Types
+
+Structure in Schema is largely expressed by example rather than by description. Types are — by default — inferred based on the values given. The Configuration Author can override these defaults using the [`@schema/type`](#schematype) annotation.
```
```suggestion
Structure in Schema is largely expressed by example rather than by description. Types are — by default — inferred based on the values given.
```
As of today (and slated for Schemas v1), setting the type explicitly is **not** supported. Only `@schema/type any=True` is supported.

pivotaljohn commented on content/ytt/docs/latest/lang-ref-ytt-schema.md:
```diff
@@ -0,0 +1,226 @@
+
+## Inferring Default Values
+
+The exact values specified in Schema are the defaults (with one notable exception — arrays — which are [detailed below](#inferring-defaults-for-arrays)). This ensures that all nodes are initialized with values, with reasonable defaults.
```
```suggestion
The exact values specified in Schema are the defaults (with one notable exception — arrays — which are [detailed below](#inferring-defaults-for-arrays)). This ensures that all nodes are properly initialized.
```
nit: slightly redundant? 🤷🏼

pivotaljohn commented on content/ytt/docs/latest/lang-ref-ytt-schema.md:
```diff
@@ -0,0 +1,226 @@
+
+### Inferring Defaults for Mappings
+
+The set of items of a map are its default: any item missing will be automatically inserted with _its_ defaults.
```
👩🏼‍🍳

pivotaljohn commented on content/ytt/docs/latest/lang-ref-ytt-schema.md:
```diff
@@ -0,0 +1,226 @@
+
+For example,
+
+`load_balancer: {}`
```
```suggestion
    ```yaml
    load_balancer: {} 
    ```
```
nit: it helps when the rendered bits are similarly formatted.

cari-lynn replied:
oof

pivotaljohn commented on content/ytt/docs/latest/lang-ref-ytt-schema.md:
```diff
@@ -0,0 +1,226 @@
+
+The default value for all arrays is an empty array. Arrays are the only type for which defaults are **not** inferred.
+
+From the example, above,
```
```suggestion
From the example, [above](#inferring-types),
```

pivotaljohn commented on content/ytt/docs/latest/lang-ref-ytt-schema.md:
```diff
@@ -0,0 +1,226 @@
+
+### @schema/type
+
+Explicitly sets the type of the annotated node.
```
This _will_ be true for the general case. However, right now, it would likely be confusing to read this sentence but see no way to give that desired type.

Perhaps we are simply clear about this, right now?

```suggestion
Explicitly configures the type of the annotated node. Currently, the only supported configuration is whether to allow any type or not.
```

pivotaljohn commented on content/ytt/docs/latest/lang-ref-ytt-schema.md:
````diff
@@ -0,0 +1,226 @@
+
+If evaluated without any data values, it results in:
+```yaml
+aws:
+  username: null
+  password: null
+
+name: null
````
```suggestion
aws: null
name: null
```

pivotaljohn commented on content/ytt/docs/latest/lang-ref-ytt-schema.md:
````diff
@@ -0,0 +1,226 @@
+name: ""
+```
+
+where `aws` and `name` are null by default. When set in a data value file `aws` will be a map with the two items, and `name` will be a string.
````
I think the example of having _no_ data values might go over well as the first part of the example... then elaborate with a concrete Data Value file and show how `aws:`, if any non-null value is given, the rest are defaulted.

cari-lynn replied:
great idea, I updated the example. Let me know if feedback

pivotaljohn reviewed (commented):
This review is focused on the main reference page.

Looking _really_ good, @cari-lynn!

Note towards the bottom, one of the examples shows the incorrect output. Details in the that comment.

cari-lynn committed:
Update content/ytt/docs/latest/ytt-schema.md

Co-authored-by: John S. Ryan <jryan@pivotal.io>

cari-lynn committed:
Update wording for adding dv files

Co-authored-by: John S. Ryan <jryan@pivotal.io>

cari-lynn committed:
Update using schema title

Co-authored-by: John S. Ryan <jryan@pivotal.io>

cari-lynn committed:
Remove use cases section

Co-authored-by: John S. Ryan <jryan@pivotal.io>

cari-lynn committed:
Order rule, then rationale

Co-authored-by: John S. Ryan <jryan@pivotal.io>

cari-lynn committed:
Leave mention of explicitly setting types for later

Co-authored-by: John S. Ryan <jryan@pivotal.io>

cari-lynn committed:
Format example

Co-authored-by: John S. Ryan <jryan@pivotal.io>

cari-lynn committed:
Link to example

Co-authored-by: John S. Ryan <jryan@pivotal.io>

cari-lynn committed:
Be clear about any type

Co-authored-by: John S. Ryan <jryan@pivotal.io>

cari-lynn committed:
correct example

Co-authored-by: John S. Ryan <jryan@pivotal.io>

cari-lynn committed:
Update defaulting language

Co-authored-by: John S. Ryan <jryan@pivotal.io>

Address feedback on..

- Separate config author files from config consumer files in example
- Simplify wording in overview of spec
- Add detail on schema overlay behavoir
- Remove unessessary wording on defaults
- Update example to show `nullable` with default values

remove sentence

Fix broken link

pivotaljohn commented on content/ytt/docs/latest/lang-ref-ytt-schema.md:
```diff
@@ -0,0 +1,244 @@
+
+### @schema/type
+
+Explicitly configures the type of the annotated node. Currently, the only supported configuration is whether to allow any type or not.
```
```suggestion
Explicitly configures the type of the annotated node. Currently, the only supported configuration is whether to allow the "any" type or not.
```

pivotaljohn commented on content/ytt/docs/latest/lang-ref-ytt-schema.md:
````diff
@@ -0,0 +1,244 @@
+@schema/type any=True
+```
+
+- `any` (`bool`) — whether or not any and all types are permitted on this node (and its children)
+    - to avoid confusion, this keyword argument is mutually exclusive to all other arguments; if this argument is
+      provided along with others, an error results.
````
```suggestion
- `any` (`bool`) — whether or not any and all types are permitted on this node (and its children).
```

There _are_ no other keyword arguments, right now (and won't be for Schema v1). Given that fact, this "note" will be confusing.

pivotaljohn commented on content/ytt/docs/latest/lang-ref-ytt-schema.md:
```diff
@@ -0,0 +1,244 @@
+
+### @schema/nullable
+
+Extends the type of the node to include "null" _and_ to be "null" by default.
```
```suggestion
Extends the type of the node to include "null" _and_ sets the default value to be `null`.
```

nit: it's the _type_ that's being set to the "null" type... and the _value_ is being set to a literal `null` value.

pivotaljohn commented on content/ytt/docs/latest/lang-ref-ytt-schema.md:
````diff
@@ -0,0 +1,244 @@
+If evaluated with the following data values:
+
+```yaml
+#@data/values
+---
+aws:
+  username: sa
+
+name: acceptance
````
What do you think of keeping `name` as `null` for this example... to illustrate how `null` _does_ flow through, if not specified... ?

```suggestion
#@data/values
---
aws:
  username: sa
```

cari-lynn replied:
good idea!

cari-lynn commented on content/ytt/docs/latest/ytt-schema.md:
```diff
@@ -0,0 +1,55 @@
+
+Consider this simple example schema and template provided by the Configuration _Author_:
+
+> `schema.yml`
```
Agreed, that seems like a good improvement.

cari-lynn commented on content/ytt/docs/latest/ytt-schema.md:
```diff
@@ -0,0 +1,55 @@
+Using a schema guarantees to templates that all data values exist and are of the correct type. This alleviates templates from doing existence and type checks themselves.
+
+
+Consider this simple example schema and template provided by the Configuration _Author_:
```
I think it's worth reverting for consistency across all docs.

Commit:
Explain how to provide non-empty array values

- remove italics
- remove indentation

cari-lynn commented on content/ytt/docs/latest/lang-ref-ytt-schema.md:
```diff
@@ -0,0 +1,244 @@
+
+- `app_domains`, if not otherwise specified is `[]`
+- `databases`, if not otherwise specified is `[]`
+
```
Good idea, I went with option two

cari-lynn committed:
UPdate mention of "any" in `schema/type` description

Co-authored-by: John S. Ryan <jryan@pivotal.io>

cari-lynn committed:
Remove suggestion of other values allowed for `schema/type`

Co-authored-by: John S. Ryan <jryan@pivotal.io>

cari-lynn committed:
Update nullable description

Co-authored-by: John S. Ryan <jryan@pivotal.io>

cari-lynn committed:
Show null as default for `nullable` in example

Co-authored-by: John S. Ryan <jryan@pivotal.io>

Commit:
Remove table of contents from "Using ytt"

- Update nullable example to have correct output

pivotaljohn commented on data/ytt/docs/ytt-latest-toc.yml:
```diff
@@ -37,6 +39,8 @@ toc:
         url: /lang-ref-ytt-version
       - page: Template module
         url: /lang-ref-ytt-template
+      - page: Schema module
```
Heads-up: _all_ modules are listed in `/lang-ref-ytt` and a subset is here.

Let's add the "Schema module" to that file as well, a la "[Overlay module](https://deploy-preview-108--carvel.netlify.app/ytt/docs/latest/lang-ref-ytt/#overlay-module)"

cari-lynn replied:
I added it at the end of the list.  :)

cari-lynn replied:
I probably belongs next to where data values are listed, however, all the modules in that section have excerpts of code there, and the schema section doesn't really have anything to put there.

Commit:
Add schema module to lang ref list
